rust/datafusion/src/execution/physical_plan/limit.rs

```diff
@@ -81,39 +82,40 @@ struct LimitPartition {
 
 impl Partition for LimitPartition {
     fn execute(&self) -> Result<Arc<Mutex<dyn RecordBatchReader + Send + Sync>>> {
-        // collect up to "limit" rows on each partition
-        let threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = self
+        // apply limit in parallel across all input partitions
+        let local_limit = self
             .partitions
             .iter()
             .map(|p| {
-                let p = p.clone();
-                let limit = self.limit;
-                thread::spawn(move || {
-                    let it = p.execute()?;
-                    collect_with_limit(it, limit)
-                })
+                Arc::new(LocalLimitExec::new(
```

Comment:
This "pick the top N from each partition and then pick the top N from the merged result" is an example of operator pushdown -- it is cool, and I wonder if there would be some way to represent this in the planner (rather than execution time).

I am thinking, for example, of pushing the limit down past `Projection`s for example.

Reply:
I agree. I would like us to move in that direction. It would be worth filing a JIRA for that as an improvement. Would you like to do that?

Reply:
will do

Reply:
Here is what I came up with -- https://issues.apache.org/jira/browse/ARROW-9732 -- I am not quite sure it captures what you had in mind